README.org
#+title: Molecular-potentials package

* Summary
  System for specifying molecular potential parameters and calculating
  molecular potentials.

  Included potentials
  - [ ] Hard Sphere
  - [X] Lennard-Jones 6-12
  - [ ] Variable hard sphere
  - [ ] Variable soft sphere
    
    
  Methods:
  - [ ] Potential calculations
  - [ ] Lennard-Jones 6-12 for collisions of two different types of molecules

* To do
  Modify format of the [[file:lennard-jones-coeffs.dat][lennard-jones-coeffs.dat]] file.  Right now it is
  based on a table from Bird & Lightfoot and contains Tc data.  I want
  to remove unnecessary data and add fields for the data source.

  


ChangeLog
2012-07-12  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* lennard-jones-coeffs.dat: Added data for SiH4 from Table 4
	Shashkov et al 2011 article.

2011-09-12  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* lennard-jones-coeffs.lisp (make-lennard-jones-6/12-potential):
	Changed call convention to accept coefficients as arguments.  A
	different routine will return an object based on a species
	(make-species-lennard-jones-6/12-potential): new routine that
	accepts the species and returns object with coefficients.

	* object-setup.lisp (molecular-potential): Removed default value
	for `epsilon/K'.  It will be set-up by individual objects

	* lennard-jones-coeffs.lisp (print-object): Same modification as
	in the hard-sphere print definition below

	* hs-molecular-potential.lisp: New 'hard-sphere-potential' class
	and helper routines.
	(print-object): Fixed insiduous bug.  I was retreiving the slot
	value of species, even if it was not bound.  This would cause an
	error that would try to print the object again, and again, and
	again ...

2011-09-11  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* object-setup.lisp (V): Modified documentation in light of new
	specification of hard-sphere potential
	(molecular-potential): Added `mass' and `species' slots.  Mass is
	important for collision calculations.  Also changed accessors from
	:accessor to :reader

	* lennard-jones-coeffs.lisp (lennard-jones-6/12-potential):
	Cleaned-up exported symbols.  From now on, other packages use the
	object interface to obtain potential parameters.

2011-09-10  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* lennard-jones-coeffs.lisp (make-lennard-jones-6/12): Removing
	the option to create a combined potential.  This is outside the
	domain of this package, but in the domain of the
	collision-integrals package.
	(lennard-jones-6/12-potential): Renamed class `lennard-jones-6/12'
	into more descriptive `lennard-jones-6/12-potential'
	(print-object, describe-object): 

2011-09-06  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* lj-molecular-potential.lisp: Renamed `lj-coeffs' into
	`lennard-jones-6/12' in the whole file

	* lennard-jones-coeffs.lisp: Renamed `lj-coeffs' into
	`lennard-jones-6/12' in the whole file

2011-09-05  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* lennard-jones-coeffs.lisp (lj-coeffs): Class inherits from new
	base class, molecular-potentials

	* generic-functions.lisp (molecular-potential): Added class
	definition of molecular-potential
	


object-setup.lisp
;; Mirko Vukovic
;; Time-stamp: <2011-09-12 13:37:39 object-setup.lisp>
;; 
;; Copyright 2011 Mirko Vukovic
;; Distributed under the terms of the GNU General Public License
;; 
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

(in-package :molecular-potentials)
(export '(V molecular-potential
	  species mass sigma epsilon/k doc))

(defgeneric V (specification r &rest rest)
  (:documentation "Calculate potential at distance `r'

Specifications is either a scalar for a hard-sphere potential or an
object that specifies the potential type (such as
hard-sphere-potential or Lennard-Jones-6/12-potential).

Currently, the hard-sphere potential can be specified by either a
scalar, or by its object.

Optional parameters can be used to specify other potential dependent
parameters"))


(defclass molecular-potential ()
  ((species :initarg :species
	    :reader species
	    :documentation "Optional string specifying the species")
   (mass :initarg :mass
	 :reader mass
	 :documentation "Species mass, in AMU")
   (sigma :initarg :sigma
	  :reader sigma
	  :documentation "Species cross-section in Angstrom")
   (epsilon/K :initarg :epsilon/K
	      :reader epsilon/K
	      :documentation "Potential-well depth, in Kelvin")
   (doc :initarg :doc
	:reader doc
	:documentation "Document data origin and other comments"))
  (:documentation "Base class for storing parameters of molecular potentials"))


(defun describe-slot (object slot stream)
  "Describe `object's `slot'

`slot' is a string whose uppercase symbol is the slot name

Output is sent to `stream'

Intended for use with `describe-object'"
  (let ((sym (symbolicate (string-upcase slot))))
    (when (slot-boundp object sym)
      (format stream "~&~a: ~a" slot (slot-value object sym)))))



hs-molecular-potential.lisp
;; Mirko Vukovic
;; Time-stamp: <2011-09-12 13:38:47 hs-molecular-potential.lisp>
;; 
;; Copyright 2011 Mirko Vukovic
;; Distributed under the terms of the GNU General Public License
;; 
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

(in-package :molecular-potentials)
(export '(hard-sphere-potential
	  make-hard-sphere-potential))

(defclass hard-sphere-potential (molecular-potential)
  ()
  (:documentation "Hard sphere potential for a species

By definition, it's epsilon/K = 0 since this potential does not have
an attractive component.

Units are NOT SI:
- mass in AMU
- sigma in Angstrom"))

(defmethod make-instance :after ((self hard-sphere-potential) &key)
  (setf (slot-value self 'epsilon/K) 0d0))

(defmethod print-object ((self hard-sphere-potential) stream)
  (print-unreadable-object (self stream :type t :identity t)
    (when (slot-boundp self 'species)
      (species self))))

(defmethod describe-object ((self hard-sphere-potential) stream)
  (format stream "Hard-sphere potential")
  (describe-slot self "Species" stream)
  (describe-slot self "Mass" stream)
  (describe-slot self "Sigma" stream))

(defun make-hard-sphere-potential (mass diameter &optional species)
  "Create object of hard-sphere potential coefficients for
`species'."
  (if species
      (make-instance 'hard-sphere-potential
		     :species species
		     :mass mass
		     :sigma diameter)
      (make-instance 'hard-sphere-potential
		     :mass mass
		     :sigma diameter)))



(defmethod V ((Radius float) r &rest rest)
  (declare (ignore rest))
  (if (> r radius)
      0.0
      most-positive-short-float))

(defmethod V ((self hard-sphere-potential) r &rest rest)
  (declare (ignore rest))
  (V (sigma self) r))

(define-test V-hs
  (assert-number-equal 0 (V pi 3.2))
  (assert-true (not (zerop (V pi 3))))
  (let ((hsp (make-hard-sphere-potential 3.2 :species :x :mass 28)))))


molecular-potentials-package-def.lisp
;; Mirko Vukovic
;; Time-stamp: <2011-09-12 09:51:08 molecular-potentials-package-def.lisp>
;; 
;; Copyright 2011 Mirko Vukovic
;; Distributed under the terms of the GNU General Public License
;; 
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

(defpackage :molecular-potentials
  (:use :cl :lisp-unit)
  (:import-from :alexandria
		:with-input-from-file
		:symbolicate)
  (:documentation "Provide objects that specify molecular potential
  parameters and methods to calculate the potentials"))

lennard-jones-molecular-potential.lisp
;; Mirko Vukovic
;; Time-stamp: <2011-09-10 23:13:52 lj-molecular-potential.lisp>
;; 
;; Copyright 2011 Mirko Vukovic
;; Distributed under the terms of the GNU General Public License
;; 
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

(in-package :molecular-potentials)

(defmethod V ((lj-coeffs lennard-jones-6/12-potential) r &rest rest)
  (declare (ignore rest))
  (with-slots (epsilon/k sigma) lj-coeffs
    (let ((sigma/r^6 (expt (/ sigma r) 6)))
    (* 4.0 epsilon/k (- (expt sigma/r^6 2)
			sigma/r^6)))))

(define-test V-LJ
  ;; verifies defining characteristics of the potential
  (with-test-coeffs
    (let ((ljc (make-lennard-jones-6/12-potential :X)))
      ;; zero crossing at r=sigma
      (assert-number-equal 0.0 (V ljc (sigma ljc)))
      ;; potential minimum = -epsilon/k at r = sigma * 2^1/6
      (assert-number-equal (- (epsilon/k ljc)) (V ljc (* (expt 2 (/ 6)) (sigma ljc)))))))

lennard-jones-coeffs.lisp
;; Mirko Vukovic
;; Time-stamp: <2011-10-20 10:06:48 lennard-jones-coeffs.lisp>
;; 
;; Copyright 2011 Mirko Vukovic
;; Distributed under the terms of the GNU General Public License
;; 
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

(in-package :molecular-potentials)

;; Lenard-Jones coefficients are used for calculating the Omega-xx
;; collision integrals.  The coefficients are read from a file and
;; stored in *lennard-jones-coeffs*.  They can be created as objects
;; using (make-lennard-jones-6/12-potential species &optional species2).
;;
;; Naming convention:
;; lj -- lennard-jones
;; coeffs -- coefficients

(export '(lennard-jones-6/12-potential
	  make-lennard-jones-6/12-potential
	  make-species-lennard-jones-6/12-potential
	  *lennard-jones-6/12-coeffs*
	  *test-lennard-jones-6/12-coeffs*))

(defparameter *data-directory*
  (asdf:system-source-directory :molecular-potentials)
  "Path to the data directory")

(defun read-data-file (file &optional (directory *data-directory*))
  "Return contents of `file' in `directory'

Signal error if file not found"
  (with-input-from-file (stream
			 (merge-pathnames file
					  directory)
			 :if-does-not-exist :error)
    (read stream)))
(defun read-lennard-jones-6/12-coeffs (&optional (file "lennard-jones-coeffs.dat"))
  "Read Lennard-Jones coefficients table"
  (read-data-file file))


(defparameter *lennard-jones-6/12-coeffs* (read-lennard-jones-6/12-coeffs)
  "table of Lennard Jones coefficients")

(defun species-lennard-jones-6/12-coeffs (species
			     &optional (db *lennard-jones-6/12-coeffs*))
  "return Lennard-Jones coefficients for `species'"
  (cdr (assoc species db)))

(defparameter *test-lennard-jones-6/12-coeffs*
  '((:X (:m 1.0) (:sigma 2.0) (:epsilon/K 4.0))
    (:Y (:m 2.0) (:sigma 3.0) (:epsilon/K 5.0)))
  "lennard-jones parameters for unit testing")

(defmacro with-test-coeffs (&body body)
  `(let* ((*lennard-jones-6/12-coeffs* *test-lennard-jones-6/12-coeffs*))
     ,@body))

(defclass lennard-jones-6/12-potential (molecular-potential)
  ()
  (:documentation "Lennard-Jones coefficients for a species

Units are NOT SI:
- mass in AMU
- sigma in Angstrom"))

(defmethod print-object ((self lennard-jones-6/12-potential) stream)
  (print-unreadable-object (self stream :type t :identity t)
    (when (slot-boundp self 'species)
      (species self))))

(defmethod describe-object ((self lennard-jones-6/12-potential) stream)
  (format stream "Lennard-Jones 6/12 potential")
  (describe-slot self "Species" stream)
  (describe-slot self "Mass" stream)
  (describe-slot self "Sigma" stream)
  (describe-slot self "epsilon/K" stream))

(defun make-lennard-jones-6/12-potential (mass sigma epsilon/K
					  &optional species)
  "Create object of Lennard-Jones 6-12 potential with coefficients
`mass', `sigma', and `epsilon/K'.  Optionally, also store the
`species' info"
  (if species
      (make-instance 'lennard-jones-6/12-potential
		     :species species
		     :mass mass
		     :sigma sigma
		     :epsilon/K epsilon/K)
      (make-instance 'lennard-jones-6/12-potential
		     :mass mass
		     :sigma sigma
		     :epsilon/K epsilon/K)))
  
(defun make-species-lennard-jones-6/12-potential (species)
  "Create object of Lennard-Jones 6-12 potential for `species'

`species' is a keyword, for a species stored in the file
*lennard-jones-6/12-coeffs* in the *data-directory*"
  (let ((coeffs (species-lennard-jones-6/12-coeffs species)))
    (make-instance 'lennard-jones-6/12-potential
		   :species species
		   :mass (cadr (assoc :m coeffs))
		   :sigma (cadr (assoc :sigma coeffs))
		   :epsilon/k (cadr (assoc :epsilon/k coeffs)))))

(define-test make-lennard-jones-6/12-potential
  (let* ((*lennard-jones-6/12-coeffs* *test-lennard-jones-6/12-coeffs*)
	 (lj-coeffs (make-lennard-jones-6/12-potential :X)))
    (assert-number-equal 1.0 (mass lj-coeffs))
    (assert-number-equal 2.0 (sigma lj-coeffs))
    (assert-number-equal 4.0 (epsilon/k lj-coeffs))))

